### ISSUE TITLE
Add Tesla Client

### ISSUE THREAD

**Issue by chris1howell** [OpenEVSE/ESP8266_WiFi_v2.x#152]
Add a Tesla client to pull Battery and temperature information.

Create Tab for vehicles

Add UI for Tesla credentials
Save Tesla Credentials to EEPROM (Alternate use credentials to get a token and discard, user would have to enter credentials periodically)
Add UI to pick vehicle (required if Tesla account has more than 1 vehicle)
Add RAPI to display battery stats to bottom line of LCD while vehicle is connected.
Add UI to allow Charge Vehicle to xxx% or xxx Miles.

```
`// TESLA Vehicle JSON API by OpenEVSE
// API Documentation https://timdorr.docs.apiary.io
//=====================================================

#include <ArduinoJson.h>
#include <ESP8266WiFi.h>
#include <WiFiClientSecure.h>

//SET Network and Tesla Account 
//=====================================================

const char* ssid = "";
const char* password = "";
const char* tesla_user = "";
const char* tesla_pass = "";

// Use web browser to view and copy
// https://www.grc.com/fingerprints.htm 
// SHA1 fingerprint of the certificate
//=====================================================

const char* fingerprint = "11:AB:EF:76:55:1F:EC:DA:31:51:61:26:97:68:FD:13:9A:04:D6:B6";
const char* tesla_host = "owner-api.teslamotors.com";
const int httpsPort = 443;
const char* TESLA_CLIENT_ID = "81527cff06843c8634fdc09e8ac0abefb46ac849f38fe1e431c2ef2106796384";
const char* TESLA_CLIENT_SECRET = "c7257eb71a564034f9419ee651c7d0e5f7aa6bfbd18bafb5c5c033b093bb2fa3";

//Static Vehicle selector until UI for choice is implemented
int vehicle = 0;

//Strings populated once authenticated with TESLA servers
String tesla_access_token = "";
String tesla_vehicle_id = "";


WiFiClientSecure client;

void setup() {
  Serial.begin(115200);
  //Serial.println();
  //Serial.print("connecting to ");
  //Serial.println(ssid);
  WiFi.begin(ssid, password);
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    //Serial.print(".");
  }
  //Serial.println("");
  //Serial.println("WiFi connected");
  //Serial.println("IP address: ");
  //Serial.println(WiFi.localIP());

  delay(1000);
  //Serial.print("connecting to ");
  //Serial.println(tesla_host);
  
  if (!client.connect(tesla_host, httpsPort)) {
    Serial.println("connection failed");
    return;
  }

  if (client.verify(fingerprint, tesla_host)) {
    //Serial.println("SSL Certificate Validated");
    if (tesla_access_token == ""){
      String s = "{";
      s += "\"grant_type\":\"password\",";
      s += "\"client_id\":\"" + String(TESLA_CLIENT_ID) + "\",";
      s += "\"client_secret\":\"" + String(TESLA_CLIENT_SECRET) + "\",";
      s += "\"email\":\"" + String(tesla_user) + "\",";
      s += "\"password\":\"" + String(tesla_pass) + "\"";
      s += "}";
      //Serial.println(s);

      client.println("POST /oauth/token HTTP/1.1");
      client.println("Host:  " + String(tesla_host));
      client.println("Connection: keep-alive");
      client.println("Content-Type: application/json; charset=utf-8");
      client.print("Content-Length: ");
      client.println(s.length());
      client.println();
      client.println(s);
      

      //Serial.print("Requesting: ");
      //Serial.println(tesla_host + String("/oauth/token"));
      while (client.connected()) {
        String json = client.readString();
        int jsonStart = json.indexOf('{');
        int jsonEnd  = (json.lastIndexOf('}') + 1);
        json = json.substring(jsonStart,jsonEnd);
        //Serial.println(json); 
        DynamicJsonBuffer jsonBuffer(512);
        JsonObject& root = jsonBuffer.parseObject(json);
        const char* access_token = root["access_token"];
        //const char* token_type = root["token_type"];
        //long expires_in = root["expires_in"];
        //const char* refresh_token = root["refresh_token"];
        //long created_at = root["created_at"];
        tesla_access_token += "Authorization: Bearer ";
        tesla_access_token += String(access_token);
        break;
        }
        client.println("GET /api/1/vehicles/ HTTP/1.1");
        client.println("Host:  " + String(tesla_host));
        client.println(tesla_access_token);
        //client.println("Connection: keep-alive");
        //client.println("Content-Length: 0");
        client.println();
        //client.println(s);

        //Serial.print("Requesting: ");
        //Serial.println(tesla_host + String("/api/1/vehicles/"));
        while (client.connected()) {
          String json = client.readString();
          int jsonStart = json.indexOf('{');
          int jsonEnd  = (json.lastIndexOf('}') + 1);
          json = json.substring(jsonStart,jsonEnd);
          //Serial.println(json);
          DynamicJsonBuffer jsonBuffer(2048);
          JsonObject& root = jsonBuffer.parseObject(json);
         
          JsonObject& response0 = root["response"][0];
          String response_id0 = response0["id"]; 
          //long response_vehicle_id0 = response0["vehicle_id"]; 
          const char* response_vin0 = response0["vin"]; 
          const char* response_display_name0 = response0["display_name"];
            //const char* response_option_codes = response["option_codes"]; 
            //const char* response_tokens0 = response["tokens"][0]; 
            //const char* response_tokens1 = response["tokens"][1]; 
            //const char* response_state = response["state"]; 
            //const char* response_id_s = response["id_s"]; 
            //bool response_remote_start_enabled = response["remote_start_enabled"]; 
            //bool response_calendar_enabled = response["calendar_enabled"]; 
            //bool response_notifications_enabled = response["notifications_enabled"]; 
          int count = root["count"]; //
          if (vehicle == 0){
            tesla_vehicle_id = response_id0;
          }
          if (count > 1){
            JsonObject& response1 = root["response"][1];
            String response_id1 = response1["id"]; 
             //long response_vehicle_id1 = response1["vehicle_id"]; 
            const char* response_vin1 = response1["vin"]; 
            const char* response_display_name1 = response1["display_name"];
            if (vehicle == 1){
              tesla_vehicle_id = response_id1;
            }  
          }
           if (count > 2){
            JsonObject& response2 = root["response"][2];
            String response_id2 = response2["id"]; 
             //long response_vehicle_id2 = response2["vehicle_id"]; 
            const char* response_vin2 = response2["vin"]; 
            const char* response_display_name2 = response2["display_name"];
            if (vehicle == 2){
              tesla_vehicle_id = response_id2;
            }  
          }          
          break;
        } 
        delay(1000);
      }
    }
  
  else {
    Serial.println("SSL Certificute Error");
  }
}

void loop() {

  
  
if (tesla_access_token != ""){
  String url = "/api/1/vehicles/" + tesla_vehicle_id + "/data_request/charge_state";
  client.println("GET " + url + " HTTP/1.1");
  client.println("Host:  " + String(tesla_host));
  client.println(tesla_access_token);
  client.println("Content-Length: 0");
  //client.println("Connection: keep-alive");
  client.println("Content-Type: application/json; charset=utf-8");
  client.println();
  //client.println(s);

  //Serial.print("Requesting: ");
  //Serial.println(tesla_host + url);
  delay(20000);
  while (client.connected()) {
    DynamicJsonBuffer jsonBuffer(2048);
    String json = client.readString();
    int jsonStart = json.indexOf('{');
    int jsonEnd  = (json.lastIndexOf('}') + 1);
    json = json.substring(jsonStart,jsonEnd);
    //Serial.println(json);
    
    JsonObject& root = jsonBuffer.parseObject(json);
    JsonObject& response = root["response"];
    const char* response_charging_state = response["charging_state"]; 
    int response_charge_limit_soc = response["charge_limit_soc"];
    //int response_charge_limit_soc_std = response["charge_limit_soc_std"]; 
    //int response_charge_limit_soc_min = response["charge_limit_soc_min"]; 
    //int response_charge_limit_soc_max = response["charge_limit_soc_max"]; 
    //bool response_charge_to_max_range = response["charge_to_max_range"]; 
    //bool response_battery_heater_on = response["battery_heater_on"]; 
    //bool response_not_enough_power_to_heat = response["not_enough_power_to_heat"]; 
    //int response_max_range_charge_counter = response["max_range_charge_counter"]; 
    //bool response_fast_charger_present = response["fast_charger_present"]; 
    //const char* response_fast_charger_type = response["fast_charger_type"]; 
    float response_battery_range = response["battery_range"]; 
    //float response_est_battery_range = response["est_battery_range"]; 
    //float response_ideal_battery_range = response["ideal_battery_range"]; 
    int response_battery_level = response["battery_level"];
    //int response_usable_battery_level = response["usable_battery_level"];
    float response_battery_current = response["battery_current"]; 
    float response_charge_energy_added = response["charge_energy_added"]; 
    float response_charge_miles_added_rated = response["charge_miles_added_rated"];
    //int response_charge_miles_added_ideal = response["charge_miles_added_ideal"];
    int response_charger_voltage = response["charger_voltage"]; 
    int response_charger_pilot_current = response["charger_pilot_current"];
    int response_charger_actual_current = response["charger_actual_current"];
    int response_charger_power = response["charger_power"]; 
    int response_time_to_full_charge = response["time_to_full_charge"]; 
    //bool response_trip_charging = response["trip_charging"]; 
    int response_charge_rate = response["charge_rate"]; 
    //bool response_charge_port_door_open = response["charge_port_door_open"]; 
    //bool response_scheduled_charging_pending = response["scheduled_charging_pending"]; 
    //bool response_charge_enable_request = response["charge_enable_request"]; 
    //const char* response_charge_port_latch = response["charge_port_latch"]; 
    //int response_charge_current_request = response["charge_current_request"];
    //int response_charge_current_request_max = response["charge_current_request_max"]; 
    //const char* response_charge_port_led_color = response["charge_port_led_color"];
    //bool response_managed_charging_active = response["managed_charging_active"]; 
    //bool response_managed_charging_user_canceled = response["managed_charging_user_canceled"];
    //bool response_motorized_charge_port = response["motorized_charge_port"]; 
    //bool response_eu_vehicle = response["eu_vehicle"]; 
    //long response_timestamp = response["timestamp"]; 

      
    for (int i=0; i <= 20; i++){
      Serial.print("$FP 0 1 ");
      Serial.print("Range - ");
      Serial.print(response_battery_range);
      Serial.println("Mi");
      delay(100);
    } 
    for (int i=0; i <= 20; i++){
      Serial.print("$FP 0 1 ");
      Serial.print("Added - ");
      Serial.print(response_charge_miles_added_rated);
      Serial.println("Mi");
      delay(100);
    }
    for (int i=0; i <= 20; i++){
      Serial.print("$FP 0 1 ");
      Serial.print("SOC Level - ");
      Serial.print(response_battery_level);
      Serial.println("% ");
      delay(100);
    } 
    for (int i=0; i <= 20; i++){
      Serial.print("$FP 0 1 ");
      Serial.print("SOC Limit - ");
      Serial.print(response_charge_limit_soc);
      Serial.println("% ");
      delay(100);
    }
    for (int i=0; i <= 20; i++){
      Serial.print("$FP 0 1 ");
      Serial.print("Done in - ");
      Serial.print(response_time_to_full_charge);
      Serial.println(" Min ");
      delay(100);
    }
    for (int i=0; i <= 20; i++){
      Serial.print("$FP 0 1 ");
      Serial.print("Voltage - ");
      Serial.print(response_charger_voltage);
      Serial.println(" VAC ");
      delay(100);
    }
    
    
    break;
  }
}



  delay(2000);
}
```

**Comment by jeremypoulter** [comment]
Is there a way to test this for those of us who don't have a Tesla?

**Comment by chris1howell** [comment]
Jeremy, contact me directly and I will supply login details. I also have a python script in addition to the ESP8266 code I posted if you want to see it...

On Sat, Feb 10, 2018 at 2:13 PM, Jeremy Poulter <notifications@github.com<mailto:notifications@github.com>> wrote:

Is there a way to test this for those of us who don't have a Tesla?

**Comment by sandeen** [comment]
It'd be pretty cool to have car stats in general; Nissan also has an API (albeit woefully slow).
I've wondered about using an ESP32 w/ bluetooth and a BT OBDII dongle to gather data directly from the car, as well.  </dream>  I guess that's a lot of scope creep.  :)

**Comment by jeremypoulter** [comment]
I wonder if it would make more sense to have the ability to pipe the car stats in via MQTT, with something with a bit more powerful interfacing with the car? While It doesn't look to challenging to implement Tesla, especially if a lot of the initial setup is done in the Web UI, I assume we will want to add support for additional cars as @sandeen eludes to. 

I guess it depends how standalone you want the system to be.

**Comment by glynhudson** [comment]
Yes, having an MQTT topic with a defined structure that the openevse wifi can subscribe to for EV specific information is a good solution I think. Since over time we will no doubt want to support an ever increasing number of EV's. It seems limited resources of the ESP does not really make it suitable for handling EV API calls. 

I agree, this would be trading off the simplicity of a standalone unit for more power elsewhere. However, it does seem that most users who would be interested in doing this are the same sort of users who are interested in diverting solar PV and other home automation / monitoring which does require an additional hub e.g. Raspberry Pi / emonPi / Synology NAS etc. . If we had a python service to post EV status to MQTT I could include this pre-installed on emonPi.

**Comment by beikeland** [comment]
Ah, there is work in progress! Its possible to do some testing against https://private-anon-9531e42980-timdorr.apiary-mock.com one downside seems to be the json is multi line from the mock server.

**Comment by chris1howell** [comment]
@beikeland The Tesla client above is fully working on the ESP8266 with the production Tesla Servers. The developers working on this are not sure this process is best handled by the client with its limited resources especially when you start adding other Manufacture/Vehicle clients. 

We are currently investigating handling this on a server and sending the data to EmonCMS. Then having the WiFi client pull down the current values through an emoncms/input/get call.

This eliminates the need for an additional MQTT device.

**Comment by beikeland** [comment]
I was perhaps unclear on why I used the term work in progress; integration to the web configuration is lacking; which is where I started browsing issues and PR's to see if I could find more details on how the html/js/c is tied together.

I would rather not have to ask someone to give me their credentials to hard code; and only ever let the code have access to the token to eliminate the possibility of starting the vehicle if exploited.

But I see there are also issues on changing the configuration so, I guess I'll teach my friend to flash this thing when he needs to update tokens. Use case is calling the `command/charge_port_door_open` api endpoint when a photo interrupter detects the charge plug has been removed from the holder.
